### ISSUE TITLE
DetachedCriteria with or (and) criteria

### ISSUE THREAD

**Issue by RNiesler** [grails/grails-data-mapping#792]
The following query will fail with NPE:
```
final detachedCriteria = new DetachedCriteria(DetachedEntity).build {
  or {
    eq 'field', 'abc'
    eq 'field', 'def'
  }
}
detachedCriteria.projections << new Query.DistinctPropertyProjection('entityId')
Entity.withCriteria {
  inList 'id', detachedCriteria
}
```
Here is the exception:
```
java.lang.NullPointerException
	at org.grails.orm.hibernate.query.AbstractHibernateQuery.doTypeConversionIfNeccessary(AbstractHibernateQuery.java:240)
	at org.grails.orm.hibernate.query.AbstractHibernateCriterionAdapter.applySubCriteriaToJunction(AbstractHibernateCriterionAdapter.java:510)
	at org.grails.orm.hibernate.query.AbstractHibernateCriterionAdapter$15.toHibernateCriterion(AbstractHibernateCriterionAdapter.java:219)
	at org.grails.orm.hibernate.query.AbstractHibernateCriterionAdapter$15.toHibernateCriterion(AbstractHibernateCriterionAdapter.java:215)
	at org.grails.orm.hibernate.query.AbstractHibernateCriterionAdapter.toHibernateCriterion(AbstractHibernateCriterionAdapter.java:531)
	at grails.orm.HibernateCriteriaBuilder.populateHibernateDetachedCriteria(HibernateCriteriaBuilder.java:225)
```
Removing the ``or`` criteria makes the query execute correctly.
Here is some code that reproduces the scenario: https://github.com/RNiesler/grails-sandbox/blob/detachedCriteria/src/integration-test/groovy/rniesler/IntegrationTest.groovy

**Comment by graemerocher** [comment]
I added a test for this and it seems to work fine with newer versions of GORM

**Comment by sudosandwich3** [comment]
I was experiencing this issue too.  To supplement what @graemerocher said, it was fixed in [this issue](https://github.com/grails/grails-data-mapping/issues/655) and released in version 5.0.2
